// elrondjs/dist/esm/common.js
/**
 * Represents the different possible types of a query result.
 */
export var ContractQueryResultDataType;
(function (ContractQueryResultDataType) {
    /**
     * Boolean value of type `Boolean`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["BOOLEAN"] = 0] = "BOOLEAN";
    /**
     * Integer of type `Number`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["INT"] = 1] = "INT";
    /**
     * Big integer of type `BigVal`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["BIG_INT"] = 2] = "BIG_INT";
    /**
     * Hex string of type `string`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["HEX"] = 3] = "HEX";
    /**
     * Generic string of type `string`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["STRING"] = 4] = "STRING";
    /**
     * Bech32 address of type `string`.
     */
    ContractQueryResultDataType[ContractQueryResultDataType["ADDRESS"] = 5] = "ADDRESS";
})(ContractQueryResultDataType || (ContractQueryResultDataType = {}));
/**
 * Transaction status.
 */
export var TransactionStatus;
(function (TransactionStatus) {
    /**
     * This means the transaction is yet to be executed by the network.
     */
    TransactionStatus[TransactionStatus["PENDING"] = 0] = "PENDING";
    /**
     * This means the transaction was executed by the network and performed all of its actions.
     */
    TransactionStatus[TransactionStatus["SUCCESS"] = 1] = "SUCCESS";
    /**
     * This means the transaction failed to be executed by the network and/or failed to perform all of its actions.
     */
    TransactionStatus[TransactionStatus["FAILURE"] = 2] = "FAILURE";
})(TransactionStatus || (TransactionStatus = {}));
